### PR TITLE
[4.0] Download key manager - GSoC Expand Extensions Manager

### PR DESCRIPTION
**Pull request by NunoLopesPT**
### Introduction

Currently there is no clear overview of which extensions have download keys to make users download their extensions with an extra_query with their keys included to allow the download. Developers have been currently hacking this system by creating their own view to add a download key to the extra_query field in the #__update_sites table (Example of an extra_query field: dlid=%s&amp;dummy=my.zip (%s is the download key)).
Each extension developer may have a different way to build this extra query so the idea is to make a view to manage and insert the download keys of all extensions in a single view in com_installer.

### Testing Instructions

Here are the extensions to test the feature:

[Component](https://github.com/NunoLopes96/component_joomla)

Plugin:
[plg_csviaddon_virtuemart_csvi_7.1.0.zip](https://github.com/joomla-projects/gsoc17_expand_extension_manager/files/1177286/plg_csviaddon_virtuemart_csvi_7.1.0.zip)

Template:
[jpeople.zip](https://github.com/joomla-projects/gsoc17_expand_extension_manager/files/1177270/jpeople.zip)

Module:
[mod_weblinks.zip](https://github.com/joomla-projects/gsoc17_expand_extension_manager/files/1177265/mod_weblinks.zip)

Package:
[pkg_weblinks_3.4.1-changelog (3).zip](https://github.com/joomla-projects/gsoc17_expand_extension_manager/files/1177262/pkg_weblinks_3.4.1-changelog.3.zip)

Library:
[lib_fof30-3.1.0-changelog (1).zip](https://github.com/joomla-projects/gsoc17_expand_extension_manager/files/1177255/lib_fof30-3.1.0-changelog.1.zip)

File:
[cli2zoombie.zip](https://github.com/joomla-projects/gsoc17_expand_extension_manager/files/1177253/cli2zoombie.zip)

To allow the prefix and sufix of the download key you need to have an <dlid> tag with prefix and sufix in your installation file (here lists.xml) like this:

`<dlid prefix="dlid=" sufix="&amp;dummy=my.zip"/>`
(note: remember to close />)

Install the component and go to Extensions -> Download Keys Manager

screenshot from 2017-07-16 17-57-50

You will not create or delete download keys so there will be only the edit button, the Download Key you see has the prefix and the suffix hidden, so the extra_query of this extension is:

`dlid=asd123asd&dummy=my.zip`

In the edit view, you will also have the prefix and sufix hidden, so the user don't have to worry about that. 
Only the developer will be responsible to define the prefix and sufix of the download key:

screenshot from 2017-07-16 18-03-32

### Modal button in com_plugins

The Downloadkeys view will probably be not known to all users, so to reinforce this new feature, a new modal button can be added to the extension when opened in com_plugins:

screenshot from 2017-07-24 18-43-23

Every plugin that has an update_site linked to it and a `<dlid>` tag on its xml installation file will pop this Download Key button in the toolbar, that will open a remote iframe to the DownloadKey view:

screenshot from 2017-07-24 18-48-40

screenshot from 2017-07-24 18-49-41

### Modal button in com_modules

The same will happen also to modules working the same way as com_plugins:
You can try this extension with an invalid updatesite and and dlid to testing porpuses:
[mod_weblinks.zip](https://github.com/joomla-projects/gsoc17_expand_extension_manager/files/1174805/mod_weblinks.zip)

screenshot from 2017-07-25 22-27-52
screenshot from 2017-07-25 22-28-23
screenshot from 2017-07-25 22-28-50

### Expected result

A simple a intuitive way to the user manage his download keys and less work to the developers to create a view to generate the extra_query

### CODE REVIEW

File: administrator/components/com_installer/View/Downloadkey/Html.php [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+	 */
+	protected function addToolbar()
+	{
+		\JFactory::getApplication()->input->set('hidemainmenu', true);
+
+		$user       = \JFactory::getUser();
+		$userId     = $user->id;
+		$checkedOut = !($this->item->checked_out == 0 || $this->item->checked_out == $userId);
+
+		// Since we don't track these assets at the item level, use the category id.
```
Comment (brianteeman) [inline administrator/components/com_installer/View/Downloadkey/Html.php]: category id?

Is that a copy paste error

File: administrator/components/com_installer/forms/filter_downloadkeys.xml [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+                description="COM_INSTALLER_UPDATESITES_FILTER_SEARCH_DESC"
+                hint="JSEARCH_FILTER"
+        />
+        <field
+                name="enabled"
+                type="list"
+                label="COM_PLUGINS_FILTER_PUBLISHED"
+                description="COM_PLUGINS_FILTER_PUBLISHED_DESC"
+                onchange="this.form.submit();"
+        >
```
Comment (brianteeman) [inline administrator/components/com_installer/forms/filter_downloadkeys.xml]: indent so the > is under the text above

File: administrator/components/com_installer/forms/filter_downloadkeys.xml [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+            <option value="">JOPTION_SELECT_PUBLISHED</option>
```
Comment (brianteeman) [inline administrator/components/com_installer/forms/filter_downloadkeys.xml]: indent

File: administrator/components/com_installer/forms/filter_downloadkeys.xml [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            <option value="0">JDISABLED</option>
+            <option value="1">JENABLED</option>
+        </field>
+        <field
+                name="client_id"
+                type="location"
+                onchange="this.form.submit();"
+        >
```
Comment (brianteeman) [inline administrator/components/com_installer/forms/filter_downloadkeys.xml]: indent one tab so the > is under the text above

**Comment by brianteeman** [comment]
Not going to comment on every xml codestyle error
Please review https://developer.joomla.org/coding-standards/xml.html

File: administrator/language/en-GB/en-GB.com_installer.ini [hunk 5]
```diff
@@ -11,6 +11,13 @@ COM_INSTALLER_CONFIRM_UNINSTALL="Are you sure you want to uninstall? Confirming
 COM_INSTALLER_CURRENT_VERSION="Installed"
 COM_INSTALLER_DISCOVER_FILTER_SEARCH_DESC="Search in discovered extension name. Prefix with ID: to search for an extension ID."
 COM_INSTALLER_DISCOVER_FILTER_SEARCH_LABEL="Search Discovered Extensions"
+COM_INSTALLER_DOWNLOADKEY_EDIT_DETAILS="Edit Download Key"
+COM_INSTALLER_DOWNLOADKEY_EDIT_EXTRA_QUERY="Download ID"
+COM_INSTALLER_DOWNLOADKEY_EDIT_TITLE="Download Key Manager"
+COM_INSTALLER_DOWNLOADKEY_EXTRA_QUERY_LABEL="Download Key"
+COM_INSTALLER_DOWNLOADKEY_EXTRA_QUERY_DESCRIPTION="Add your Download key"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_installer.ini]: As this is a description please add a period at the end.

File: administrator/language/en-GB/en-GB.com_installer.ini [hunk 6]
```diff
@@ -34,10 +41,12 @@ COM_INSTALLER_HEADER_LANGUAGES="Extensions: Install Languages"
 COM_INSTALLER_HEADER_MANAGE="Extensions: Manage"
 COM_INSTALLER_HEADER_UPDATE="Extensions: Update"
 COM_INSTALLER_HEADER_UPDATESITES="Extensions: Update Sites"
+COM_INSTALLER_HEADER_DOWNLOADKEYS="Extensions: Download Keys Manager"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_installer.ini]: Please dont use the word Manager anywhere - we went to a lot of trouble to remove that word

File: administrator/language/en-GB/en-GB.ini [hunk 7]
```diff
@@ -336,6 +336,8 @@ JGLOBAL_DISPLAY_NUM="Display #"
 JGLOBAL_DISPLAY_SELECT_DESC="Show or hide the Display Select dropdown listbox."
 JGLOBAL_DISPLAY_SELECT_LABEL="Display Select"
 JGLOBAL_DOWN="Down"
+JGLOBAL_DOWNLOAD_KEY_MODAL_BUTTON="Download key"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.ini]: Please change to Download Key for consistency

File: administrator/components/com_installer/tmpl/downloadkeys/default.php [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+								</td>
+							</tr>
+						</tfoot>
+						<tbody>
+						<?php $itemsSkipped = 0; ?>
+						<?php foreach ($this->items as $i => $item) : ?>
+							<?php if ($item->extra_query == null): ?>
+								<?php $itemsSkipped++; ?>
+								<?php continue; ?>
+							<?php endif;?>
```
Comment (C-Lodder) [inline administrator/components/com_installer/tmpl/downloadkeys/default.php]: space after `;`

File: administrator/components/com_installer/tmpl/downloadkey/edit.php [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+?>
+<form action="<?php echo JRoute::_('index.php?option=com_installer&view=downloadkey&layout=edit&update_site_id=' . (int) $this->item->update_site_id . $tmp); ?>"
+	  method="post" name="adminForm" id="adminForm">
+	<div class="form-horizontal">
+		<fieldset class="adminform">
+			<legend><?php echo JText::_('COM_INSTALLER_DOWNLOADKEY_EDIT_DETAILS'); ?></legend>
+			<div class="row">
+				<div class="col">
+					<?php foreach ($this->form->getFieldset() as $field): ?>
```
Comment (C-Lodder) [inline administrator/components/com_installer/tmpl/downloadkey/edit.php]: space before `:`

File: administrator/components/com_installer/tmpl/downloadkeys/default.php [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+							<tr>
+								<td colspan="8">
+									<?php echo $this->pagination->getListFooter(); ?>
+								</td>
+							</tr>
+						</tfoot>
+						<tbody>
+						<?php $itemsSkipped = 0; ?>
+						<?php foreach ($this->items as $i => $item) : ?>
+							<?php if ($item->extra_query == null): ?>
```
Comment (C-Lodder) [inline administrator/components/com_installer/tmpl/downloadkeys/default.php]: space before `:`

**Comment by brianteeman** [comment]
Can you look at resolving the conflicts so that this can be tested please.

**Comment by roland-d** [comment]
@wilsonge I started refactoring this for J4.

**Comment by roland-d** [comment]
Closing this in favour of #25553

**Comment by ka3media** [comment]
As far as I can see the keys will be shown and stored in the clear.

From my practical experience there are several situations where you use your own key for client pages or nonsalaried work (for example for a club). And you are not always able to hide it using user access.

This will be a problem with normal keys, but a critical problem if those keys are related to 
volume-based billing, e.g picture optimizing.

Is there a way to hash the keys or or at least hide them in the backend?

**Comment by brianteeman** [comment]
@ka3media please create a new issue for your question. Closed issues are rarely seen

**Comment by roland-d** [comment]
@ka3media Perhaps add it to this issue: https://issues.joomla.org/tracker/joomla-cms/26583

**Pull request closed without merging**